### PR TITLE
[ROCM] GroupedGemm 2d2d has potential uninitalized data

### PR DESCRIPTION
**Pull request by alex-minooka**
Summary

Torchtitan upstream on ROCM using grouped_gemm was seeing gradient norm going to NaN after small number of training steps. Issue occurred when experts were given small <128 number of tokens. 

Change:
Modifying rocm grouped gemm to initialize to zeros if given 2d2d shape and to skip expert if expert has no tokens. After code changes, grad nan issue in torchtitan upstream is resolved

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang

### CODE REVIEW

**Comment by jagadish-amd** [comment]
> Summary
> 
> Torchtitan upstream on ROCM using scaled_grouped_gemm was seeing gradient norm going to NaN after small number of training steps. Issue occurred when experts were given small <128 number of tokens.
> 
> Change: Modifying rocm grouped gemm to initialize to zeros if given 2d2d shape and to skip expert if expert is no tokens. After code changes, grad nan issue in torchtitan upstream is resolved
> 
nitpick
scaled_grouped_gemm --> should be grouped_gemm
skip expert if expert is no tokens --> skip expert if expert has no tokens.

**Comment by alex-minooka** [comment]
> @alex-minooka Thanks for the PR. Current test cases are at test_matmul_cuda.py PYTORCH_TEST_WITH_ROCM=1 pytest test/test_matmul_cuda.py -v -k "test_grouped_gemm_2d_2d or test_grouped_gemm_2d_3d or test_grouped_gemm_3d_3d or test_grouped_gemm_3d_2d" To test with CK backend, ROCM_ALLOW_GROUP_GEMM_CK=1 PYTORCH_TEST_WITH_ROCM=1 pytest test/test_matmul_cuda.py -v -k "test_grouped_gemm_2d_2d or test_grouped_gemm_2d_3d or test_grouped_gemm_3d_3d or test_grouped_gemm_3d_2d" Maybe we need to add a test case to check this bug. cc @jeffdaily

PR passes both of these test cases.

**Review by jeffdaily — APPROVED** [review]
